std/assembly/array.ts

```diff
@@ -476,6 +476,70 @@ export class Array<T> extends ArrayBufferView {
     return this;
   }
 
+  flat(): valueof<T>[] {
+    if (!isArray<T>()) {
+      ERROR("Cannot call flat() on Array<T> where T is not an Array.");
+    }
+    // Get the length and data start values
+    let length = this.length_;
+    let selfDataStart = this.dataStart;
+
+    // calculate the end size with an initial pass
+    let size: i32 = 0;
+    for (let i = 0; i < length; i++) {
+      let child: T = load<T>(selfDataStart + (i << alignof<T>()));
+      size += child == null ? 0 : child.length;
+    }
+
+    // calculate the byteLength of the resulting backing ArrayBuffer
+    let byteLength = size << usize(alignof<valueof<T>>());
+    let dataStart = __alloc(byteLength, idof<ArrayBuffer>());
+
+    // create the return value
+    let result: valueof<T>[] = changetype<valueof<T>[]>(__alloc(offsetof<valueof<T>[]>(), idof<valueof<T>[]>()));
+
+    // This store is required, because the `length_` property is private
+    store<i32>(changetype<usize>(result), size, offsetof<valueof<T>[]>("length_"));
+
+    // Set the dataLength and dataStart properties
+    result.dataLength = byteLength;
+    result.dataStart = dataStart;
+    // this statement retains the ArrayBuffer
+    result.data = changetype<ArrayBuffer>(dataStart);
+
+    // set the elements
+    let resultIndex: i32 = -1;
+    for (let i = 0; i < length; i++) { // for each child
+      let child: T = load<T>(selfDataStart + (i << alignof<T>()));
+
+      // ignore null arrays
+      if (child == null) continue;
```

Comment:
Does this maybe miss zeroing the element in the backing buffer? Above `__alloc` is used to make it, which does not zero.

Reply:
All the elements will be set so using an alloc without zeroing the data is safe here.

Reply:
Ah, right, it skips the element entirely.